File: backend/app/api/evaluations.py
"""Agent 评测中心 owner-scoped HTTP API。"""

from __future__ import annotations

from datetime import datetime
from typing import NoReturn

from ai.workflows.evaluation import EvaluationUseCaseError, evaluation_use_cases
from app.api.deps import get_current_user_id
from app.schemas.evaluations import (
    EvaluationAdjudicationRequest,
    EvaluationAnnotationCreateRequest,
    EvaluationCalibrationCreateRequest,
    EvaluationCalibrationSimulateRequest,
    EvaluationCandidateDatasetRequest,
    EvaluationDatasetCreateRequest,
    EvaluationDatasetStatusRequest,
    EvaluationGatePolicyCreateRequest,
    EvaluationOnlineSampleRequest,
    EvaluationReviewRequest,
    EvaluationRunCreateRequest,
    EvaluationSuiteCreateRequest,
)
from fastapi import APIRouter, Depends, Header, HTTPException, Query
from fastapi.responses import HTMLResponse

router = APIRouter(prefix="/api/evaluations", tags=["Agent 评测"])


def _raise(exc: EvaluationUseCaseError) -> NoReturn:
    """将应用层错误映射为稳定 HTTP 响应。"""

    raise HTTPException(status_code=exc.status_code, detail=exc.message) from exc


@router.get("/overview")
async def overview(user_id: str = Depends(get_current_user_id)):
    """返回运行、语义、完全成功和人工队列总览。"""

    try:
        return await evaluation_use_cases.overview(user_id=user_id)
    except EvaluationUseCaseError as exc:
        _raise(exc)


@router.get("/metrics/trends")
async def trends(
    agent_name: str | None = Query(default=None),
    agent_version: str | None = Query(default=None),
    prompt_name: str | None = Query(default=None),
    prompt_version: str | None = Query(default=None),
    model_config_hash: str | None = Query(default=None),
    dataset_version: str | None = Query(default=None),
    environment: str | None = Query(default=None),
    created_from: datetime | None = Query(default=None),
    created_to: datetime | None = Query(default=None),
    user_id: str = Depends(get_current_user_id),
):
    """返回版本质量、延迟和 Token 趋势。"""

    try:
        return await evaluation_use_cases.trends(
            user_id=user_id,
            agent_name=agent_name,
            agent_version=agent_version,
            prompt_name=prompt_name,
            prompt_version=prompt_version,
            model_config_hash=model_config_hash,
            dataset_version=dataset_version,
            environment=environment,
            created_from=created_from,
            created_to=created_to,
        )
    except EvaluationUseCaseError as exc:
        _raise(exc)


@router.get("/regressions")
async def regressions(user_id: str = Depends(get_current_user_id)):
    """返回确认的回归告警。"""

    try:
        return await evaluation_use_cases.regressions(user_id=user_id)
    except EvaluationUseCaseError as exc:
        _raise(exc)


@router.get("/suites")
async def list_suites(user_id: str = Depends(get_current_user_id)):
    """列出当前用户评测套件。"""

    try:
        return await evaluation_use_cases.list_suites(user_id=user_id)
    except EvaluationUseCaseError as exc:
        _raise(exc)


@router.post("/suites", status_code=201)
async def create_suite(
    request: EvaluationSuiteCreateRequest,
    user_id: str = Depends(get_current_user_id),
):
    """创建评测套件。"""

    try:
        return await evaluation_use_cases.create_suite(user_id=user_id, request=request)
    except EvaluationUseCaseError as exc:
        _raise(exc)


@router.get("/datasets")
async def list_datasets(
    limit: int = Query(default=50, ge=1, le=200),
    offset: int = Query(default=0, ge=0),
    user_id: str = Depends(get_current_user_id),
):
    """分页返回数据集元数据。"""

    try:
        return await evaluation_use_cases.list_datasets(
            user_id=user_id, limit=limit, offset=offset
        )
    except EvaluationUseCaseError as exc:
        _raise(exc)


@router.post("/datasets", status_code=201)
async def create_dataset(
    request: EvaluationDatasetCreateRequest,
    user_id: str = Depends(get_current_user_id),
):
    """创建包含加密案例的数据集版本。"""

    try:
        return await evaluation_use_cases.create_dataset(user_id=user_id, request=request)
    except EvaluationUseCaseError as exc:
        _raise(exc)


@router.get("/datasets/{dataset_id}")
async def get_dataset(dataset_id: str, user_id: str = Depends(get_current_user_id)):
    """获取数据集安全元数据。"""

    try:
        return await evaluation_use_cases.get_dataset(
            user_id=user_id, dataset_id=dataset_id
        )
    except EvaluationUseCaseError as exc:
        _raise(exc)


@router.post("/datasets/{dataset_id}/lock")
async def lock_dataset(dataset_id: str, user_id: str = Depends(get_current_user_id)):
    """锁定 calibrated 数据集版本。"""

    try:
        return await evaluation_use_cases.lock_dataset(
            user_id=user_id, dataset_id=dataset_id
        )
    except EvaluationUseCaseError as exc:
        _raise(exc)


@router.post("/datasets/{dataset_id}/status")
async def update_dataset_status(
    dataset_id: str,
    request: EvaluationDatasetStatusRequest,
    user_id: str = Depends(get_current_user_id),
):
    """推进 Dataset Version 生命周期，不允许回退或修改已锁定版本。"""

    try:
        return await evaluation_use_cases.update_dataset_status(
            user_id=user_id, dataset_id=dataset_id, request=request
        )
    except EvaluationUseCaseError as exc:
        _raise(exc)


@router.post("/runs", status_code=202)
async def create_run(
    request: EvaluationRunCreateRequest,
    user_id: str = Depends(get_current_user_id),
    idempotency_key: str | None = Header(default=None, alias="Idempotency-Key"),
):
    """创建有预算和并发上限的可恢复评测任务。"""

    try:
        return await evaluation_use_cases.create_run(
            user_id=user_id, request=request, idempotency_key=idempotency_key
        )
    except EvaluationUseCaseError as exc:
        _raise(exc)


@router.get("/runs")
async def list_runs(
    limit: int = Query(default=50, ge=1, le=200),
    offset: int = Query(default=0, ge=0),
    user_id: str = Depends(get_current_user_id),
):
    """分页列出评测运行。"""

    try:
        return await evaluation_use_cases.list_runs(
            user_id=user_id, limit=limit, offset=offset
        )
    except EvaluationUseCaseError as exc:
        _raise(exc)


@router.get("/runs/{run_id}")
async def get_run(run_id: str, user_id: str = Depends(get_current_user_id)):
    """获取评测运行和案例摘要。"""

    try:
        return await evaluation_use_cases.get_run(user_id=user_id, run_id=run_id)
    except EvaluationUseCaseError as exc:
        _raise(exc)


@router.post("/runs/{run_id}/cancel")
async def cancel_run(run_id: str, user_id: str = Depends(get_current_user_id)):
    """取消关联 AgentRun。"""

    try:
        return await evaluation_use_cases.cancel_run(user_id=user_id, run_id=run_id)
    except EvaluationUseCaseError as exc:
        _raise(exc)


@router.post("/runs/{run_id}/retry-failed")
async def retry_failed(run_id: str, user_id: str = Depends(get_current_user_id)):
    """重试失败评测任务。"""

    try:
        return await evaluation_use_cases.retry_failed(user_id=user_id, run_id=run_id)
    except EvaluationUseCaseError as exc:
        _raise(exc)


@router.post("/runs/{run_id}/request-review")
async def request_review(
    run_id: str,
    request: EvaluationReviewRequest,
    user_id: str = Depends(get_current_user_id),
):
    """把失败案例或显式选中的案例加入人工复核队列。"""

    try:
        return await evaluation_use_cases.request_review(
            user_id=user_id, run_id=run_id, request=request
        )
    except EvaluationUseCaseError as exc:
        _raise(exc)


@router.get("/runs/{run_id}/events")
async def events(
    run_id: str,
    after_sequence: int = Query(default=0, ge=0),
    limit: int = Query(default=200, ge=1, le=500),
    user_id: str = Depends(get_current_user_id),
):
    """返回可重放 AgentRun 事件。"""

    try:
        return await evaluation_use_cases.events(
            user_id=user_id,
            run_id=run_id,
            after_sequence=after_sequence,
            limit=limit,
        )
    except EvaluationUseCaseError as exc:
        _raise(exc)


@router.get("/runs/{run_id}/report")
async def export_report(
    run_id: str,
    format: str = Query(default="json", pattern=r"^(json|html)$"),
    user_id: str = Depends(get_current_user_id),
):
    """导出 JSON 或无脚本 HTML 评测报告。"""

    try:
        report = await evaluation_use_cases.export_report(
            user_id=user_id, run_id=run_id, output_format=format
        )
        if format == "html":
            return HTMLResponse(content=str(report))
        return report
    except EvaluationUseCaseError as exc:
        _raise(exc)


@router.get("/runs/{run_id}/cases")
async def list_case_runs(run_id: str, user_id: str = Depends(get_current_user_id)):
    """列出运行内案例摘要。"""

    try:
        return await evaluation_use_cases.list_case_runs(
            user_id=user_id, run_id=run_id
        )
    except EvaluationUseCaseError as exc:
        _raise(exc)


@router.get("/case-runs/{case_run_id}")
async def get_case_run(
    case_run_id: str, user_id: str = Depends(get_current_user_id)
):
    """按 owner 加载一个案例的实际输出和脱敏轨迹。"""

    try:
        return await evaluation_use_cases.get_case_run(
            user_id=user_id, case_run_id=case_run_id
        )
    except EvaluationUseCaseError as exc:
        _raise(exc)


@router.get("/annotations/queue")
async def annotation_queue(user_id: str = Depends(get_current_user_id)):
    """返回 needs_review 人工队列。"""

    try:
        return await evaluation_use_cases.annotation_queue(user_id=user_id)
    except EvaluationUseCaseError as exc:
        _raise(exc)


@router.post("/case-runs/{case_run_id}/annotations", status_code=201)
async def add_annotation(
    case_run_id: str,
    request: EvaluationAnnotationCreateRequest,
    user_id: str = Depends(get_current_user_id),
):
    """追加人工标注 revision。"""

    try:
        return await evaluation_use_cases.add_annotation(
            user_id=user_id, case_run_id=case_run_id, request=request
        )
    except EvaluationUseCaseError as exc:
        _raise(exc)


@router.post("/case-runs/{case_run_id}/candidate-dataset", status_code=201)
async def create_candidate_dataset(
    case_run_id: str,
    request: EvaluationCandidateDatasetRequest,
    user_id: str = Depends(get_current_user_id),
):
    """把人工确认的失败案例沉淀为新的 Dataset Version。"""

    try:
        return await evaluation_use_cases.create_candidate_dataset(
            user_id=user_id, case_run_id=case_run_id, request=request
        )
    except EvaluationUseCaseError as exc:
        _raise(exc)


@router.get("/case-runs/{case_run_id}/annotations")
async def list_annotations(
    case_run_id: str, user_id: str = Depends(get_current_user_id)
):
    """返回案例标注历史。"""

    try:
        return await evaluation_use_cases.list_annotations(
            user_id=user_id, case_run_id=case_run_id
        )
    except EvaluationUseCaseError as exc:
        _raise(exc)


@router.post("/annotations/{annotation_id}/adjudicate", status_code=201)
async def adjudicate(
    annotation_id: str,
    request: EvaluationAdjudicationRequest,
    user_id: str = Depends(get_current_user_id),
):
    """追加专家裁决 revision。"""

    try:
        return await evaluation_use_cases.adjudicate(
            user_id=user_id, annotation_id=annotation_id, request=request
        )
    except EvaluationUseCaseError as exc:
        _raise(exc)


@router.get("/calibrations")
async def list_calibrations(user_id: str = Depends(get_current_user_id)):
    """列出 Judge Calibration Versions。"""

    try:
        return await evaluation_use_cases.list_calibrations(user_id=user_id)
    except EvaluationUseCaseError as exc:
        _raise(exc)


@router.post("/calibrations", status_code=201)
async def create_calibration(
    request: EvaluationCalibrationCreateRequest,
    user_id: str = Depends(get_current_user_id),
):
    """创建不可变 Judge Calibration Version。"""

    try:
        return await evaluation_use_cases.create_calibration(
            user_id=user_id, request=request
        )
    except EvaluationUseCaseError as exc:
        _raise(exc)


@router.post("/calibrations/{calibration_id}/simulate")
async def simulate_calibration(
    calibration_id: str,
    request: EvaluationCalibrationSimulateRequest,
    user_id: str = Depends(get_current_user_id),
):
    """模拟阈值变化；calibration_id 仅用于前端上下文，不修改历史版本。"""

    try:
        return await evaluation_use_cases.simulate_calibration(
            user_id=user_id,
            calibration_id=calibration_id,
            request=request,
        )
    except EvaluationUseCaseError as exc:
        _raise(exc)


@router.post("/online-samples/evaluate")
async def evaluate_online_sample(
    request: EvaluationOnlineSampleRequest,
    _user_id: str = Depends(get_current_user_id),
):
    """脱敏生产 Trace，并执行风险分层的抽样决策。"""

    try:
        return evaluation_use_cases.online_sample(request=request)
    except EvaluationUseCaseError as exc:
        _raise(exc)


@router.get("/gates")
async def list_gates(user_id: str = Depends(get_current_user_id)):
    """列出 Gate Policy Versions。"""

    try:
        return await evaluation_use_cases.list_gate_policies(user_id=user_id)
    except EvaluationUseCaseError as exc:
        _raise(exc)


@router.post("/gates", status_code=201)
async def create_gate(
    request: EvaluationGatePolicyCreateRequest,
    user_id: str = Depends(get_current_user_id),
):
    """创建 Gate Policy Version。"""

    try:
        return await evaluation_use_cases.create_gate_policy(
            user_id=user_id, request=request
        )
    except EvaluationUseCaseError as exc:
        _raise(exc)


@router.post("/runs/{run_id}/gate-check")
async def gate_check(
    run_id: str,
    policy_id: str | None = Query(default=None),
    user_id: str = Depends(get_current_user_id),
):
    """执行发布门禁并保存不可变结果。"""

    try:
        return await evaluation_use_cases.gate_check(
            user_id=user_id, run_id=run_id, policy_id=policy_id
        )
    except EvaluationUseCaseError as exc:
        _raise(exc)

File: web/lib/api/evaluations.ts
import { apiRequest, authFetch } from './config';

export type EvaluationAgentName =
    | 'interview_planner'
    | 'interview_turn'
    | 'interview_scoring'
    | 'resume_optimizer'
    | 'resume_analyzer';

export type EvaluationQuickModeName = 'quick' | 'standard' | 'release';

export interface EvaluationCatalogAgent {
    name: EvaluationAgentName;
    label: string;
    description: string;
    prompt_name: string;
    prompt_version: string;
    dataset_name: string;
    dataset_version: string;
    suite_name: string;
    rubric_version: string;
    case_count: number;
    latest_successful_run_id: string | null;
}

export interface EvaluationQuickMode {
    name: EvaluationQuickModeName;
    label: string;
    description: string;
    repetition_count: number;
    max_concurrency: number;
    max_budget_usd: number;
    max_cases: number | null;
    include_judges: boolean;
    human_review_rate: number;
}

export interface EvaluationCatalog {
    agents: EvaluationCatalogAgent[];
    modes: EvaluationQuickMode[];
    runs_enabled: boolean;
}

export interface EvaluationQuickRunRequest {
    agent_name: EvaluationAgentName;
    mode: EvaluationQuickModeName;
    /** Request-scoped credentials; callers must never log or render this object. */
    api_config: Record<string, unknown>;
    prompt_name?: string;
    prompt_version?: string;
    compare_production?: boolean;
}

export interface EvaluationOverview {
    run_count: number;
    runtime_success_rate: number | null;
    semantic_success_rate: number | null;
    complete_success_rate: number | null;
    hard_gate_pass_rate: number | null;
    factual_support_rate: number | null;
    tool_call_accuracy: number | null;
    judge_human_agreement: number | null;
    pending_review_count: number;
    regression_count: number;
    p95_latency_ms: number | null;
    token_delta_percent: number | null;
    trace_completeness_rate: number | null;
    trace_incomplete_count: number;
    tool_failure_rate: number | null;
    tool_execution_success_rate: number | null;
    tool_p95_duration_ms: number | null;
    dependency_failure_rate: number | null;
    external_io_timeout_rate: number | null;
    retrieval_empty_rate: number | null;
    external_effect_count: number;
    external_effect_blocked_count: number;
    approval_event_count: number;
    approval_violation_count: number;
    langfuse_reported_case_count: number;
    langfuse_failed_case_count: number;
}

export interface EvaluationDatasetCase {
    id: string;
    case_key: string;
    category: string;
    tags: string[];
    severity: 'low' | 'medium' | 'high' | 'critical';
    content_hash: string;
    created_at: string;
}

export interface EvaluationDataset {
    id: string;
    name: string;
    version: string;
    status: string;
    case_count: number;
    source: string;
    content_hash: string;
    created_at: string;
    locked_at: string | null;
}

export interface EvaluationDatasetDetail extends EvaluationDataset {
    cases: EvaluationDatasetCase[];
}

export interface EvaluationSuite {
    id: string;
    name: string;
    agent_name: string;
    description: string | null;
    dataset_version_id: string;
    rubric_version: string;
    gate_policy_id: string | null;
}

export interface EvaluationRunSummary {
    case_total?: number;
    completed_count?: number;
    failed_count?: number;
    hard_gate_failure_count?: number;
    needs_review_count?: number;
    progress?: number;
    complete_success_rate?: number | null;
    p95_latency_ms?: number | null;
    token_total?: number | null;
    trace_complete_count?: number;
    trace_incomplete_count?: number;
    trace_completeness_rate?: number | null;
    tool_call_total?: number;
    tool_call_completed_count?: number;
    tool_call_failed_count?: number;
    tool_failure_rate?: number | null;
    tool_execution_success_rate?: number | null;
    tool_p95_duration_ms?: number | null;
    external_effect_total?: number;
    external_effect_blocked_count?: number;
    approval_violation_count?: number;
    external_io_total?: number;
    external_io_failed_count?: number;
    external_io_timeout_count?: number;
    dependency_failure_rate?: number | null;
    external_io_timeout_rate?: number | null;
    approval_event_total?: number;
    retrieval_observed_case_count?: number;
    retrieval_empty_case_count?: number;
    retrieval_empty_rate?: number | null;
    langfuse_reported_case_count?: number;
    langfuse_failed_case_count?: number;
    [key: string]: unknown;
}

export interface EvaluationRun {
    id: string;
    suite_id: string;
    agent_run_id: string | null;
    agent_name: string;
    agent_version: string;
    prompt_name: string | null;
    prompt_version: string | null;
    model_config_hash: string;
    dataset_version: string;
    status: string;
    baseline_run_id: string | null;
    repetition_count: number;
    include_judges: boolean;
    budget: Record<string, unknown>;
    summary: EvaluationRunSummary;
    created_at: string;
    started_at: string | null;
    finished_at: string | null;
}

export interface EvaluationScore {
    id: string;
    case_run_id: string;
    metric_name: string;
    value: number | null;
    status: string;
    source: 'deterministic' | 'deepeval' | 'judge' | 'human' | 'user_feedback' | string;
    reason: string | null;
    severity: string;
    hard_gate: boolean;
    evidence_refs: string[];
    metric_version: string;
    created_at: string;
}

export interface EvaluationCaseRun {
    id: string;
    evaluation_run_id: string;
    case_id: string;
    repetition_index: number;
    status: string;
    trace_id: string | null;
    latency_ms: number;
    token_usage: Record<string, number>;
    hard_gate_passed: boolean;
    overall_score: number | null;
    error_category: string | null;
    needs_review: boolean;
}

export type EvaluationToolEffect = 'none' | 'read' | 'write' | 'external';
export type EvaluationToolStatus = 'requested' | 'started' | 'completed' | 'failed' | 'blocked' | 'skipped' | string;
export type EvaluationApprovalStatus = 'not_required' | 'pending' | 'approved' | 'rejected' | string;

export interface EvaluationToolCall {
    call_id: string;
    sequence: number;
    event_type?: string;
    parent_call_id?: string | null;
    tool_name: string;
    effect: EvaluationToolEffect;
    status: EvaluationToolStatus;
    attempt: number;
    requires_confirmation: boolean;
    approval_status: EvaluationApprovalStatus;
    simulated: boolean;
    duration_ms?: number | null;
    error_type?: string | null;
    error_category?: string | null;
    evidence_refs: string[];
}

export interface EvaluationRetrieval {
    retrieval_id: string;
    sequence?: number | null;
    source_type: string;
    strategy?: string | null;
    result_count?: number | null;
    empty_result?: boolean | null;
    adopted: boolean;
    duration_ms?: number | null;
    error_category?: string | null;
}

export interface EvaluationExternalIo {
    call_id: string;
    sequence: number;
    event_type?: string;
    parent_call_id?: string | null;
    operation: string;
    dependency?: string | null;
    status: string;
    attempt: number;
    duration_ms?: number | null;
    item_count?: number | null;
    result_count?: number | null;
    adopted?: boolean | null;
    error_type?: string | null;
    error_category?: string | null;
}

export interface EvaluationModelCall {
    call_id: string;
    sequence: number;
    model_channel: string;
    model_member_hash: string;
    fallback_index: number;
    latency_ms: number;
    status: string;
    error_classification?: string | null;
}

export interface EvaluationApproval {
    approval_id: string;
    action: string;
    status: EvaluationApprovalStatus;
    call_id?: string | null;
    requested_sequence?: number | null;
    sequence?: number | null;
    evidence_refs: string[];
}

export interface EvaluationRunEvent {
    sequence: number;
    stage: string;
    event_type: string;
    status?: string | null;
    payload_summary: Record<string, unknown>;
}

export interface EvaluationTraceCompleteness {
    complete: boolean;
    score: number;
    missing: string[];
    trace_id_present: boolean;
    tool_terminal_states_complete: boolean;
    stable_error_categories: boolean;
    external_approval_status_present: boolean;
    agent_run_id_present: boolean;
}

export interface EvaluationObservabilitySummary {
    schema_version: number;
    tool_event_count: number;
    tool_call_summary: Record<string, unknown>;
    external_io_event_count: number;
    external_io_summary: Record<string, unknown>;
    approval_event_count: number;
    langfuse_reported?: boolean | null;
    langfuse_error?: string | null;
    trace_completeness: EvaluationTraceCompleteness;
}

export interface EvaluationRecord {
    final_status?: string;
    trace_id?: string | null;
    agent_run_id?: string | null;
    tool_calls?: EvaluationToolCall[];
    retrievals?: EvaluationRetrieval[];
    external_ios?: EvaluationExternalIo[];
    model_calls?: EvaluationModelCall[];
    approvals?: EvaluationApproval[];
    events?: EvaluationRunEvent[];
    observability?: EvaluationObservabilitySummary;
    recovery_count?: number;
    estimated_cost_usd?: number | null;
    error?: { classification?: string; message?: string; retryable?: boolean } | null;
    [key: string]: unknown;
}

export interface EvaluationCaseRunDetail extends EvaluationCaseRun {
    actual_output: unknown;
    pairwise_outputs: { A: unknown; B: unknown | null };
    record: EvaluationRecord;
    case: EvaluationDatasetCase & {
        input: Record<string, unknown>;
        expected: Record<string, unknown>;
    };
    scores: EvaluationScore[];
    annotations: EvaluationAnnotation[];
}

export interface EvaluationAnnotation {
    id: string;
    case_run_id: string;
    rubric_version: string;
    annotation_type: string;
    metric_name: string;
    value: unknown;
    labels: string[];
    evidence_spans: Array<{ start: number; end: number; text: string }>;
    comment: string | null;
    confidence: number | null;
    reviewer_key: string;
    blind: boolean;
    revision: number;
    adjudication: boolean;
    created_at: string;
}

export interface EvaluationCalibration {
    id: string;
    metric_name: string;
    judge_version: string;
    dataset_version: string;
    human_sample_count: number;
    statistics: Record<string, number | null>;
    threshold: number | null;
    status: string;
    created_at: string;
}

export interface EvaluationGatePolicy {
    id: string;
    name: string;
    version: string;
    hard_gates: string[];
    metric_thresholds: Record<string, number | { value: number; comparison: 'gte' | 'lte' | 'eq' }>;
    regression_tolerances: Record<string, unknown>;
    minimum_sample_size: number;
    status: string;
    created_at: string;
}

export interface EvaluationTrendPoint {
    run_id: string;
    created_at: string;
    environment: string;
    agent_name: string;
    agent_version: string;
    prompt_name: string | null;
    prompt_version: string | null;
    model_config_hash: string;
    dataset_version: string;
    average_score: number | null;
    minimum_score: number | null;
    score_spread: number | null;
    sample_count: number;
    complete_success_rate: number | null;
    p50_latency_ms: number | null;
    p95_latency_ms: number | null;
    token_total: number | null;
    p50_tokens: number | null;
    p95_tokens: number | null;
}

export interface EvaluationRegression {
    run_id: string;
    baseline_run_id: string | null;
    agent_name: string;
    agent_version: string;
    prompt_name: string | null;
    prompt_version: string | null;
    model_config_hash: string;
    dataset_version: string;
    severity: string;
    regression_count: number;
    failed_case_count: number;
    hard_gate_failure_count: number;
    hard_gate_blocked: boolean;
    metric_deltas: Record<string, number>;
    complete_success_rate_delta: number | null;
    p95_latency_ms_delta: number | null;
    created_at: string;
}

export interface EvaluationTrendFilters {
    agent_name?: string;
    agent_version?: string;
    prompt_name?: string;
    prompt_version?: string;
    model_config_hash?: string;
    dataset_version?: string;
    environment?: string;
    created_from?: string;
    created_to?: string;
}

export interface EvaluationCaseRunFilters {
    status?: string;
    error_category?: string;
    tool_name?: string;
    tool_effect?: EvaluationToolEffect;
    tool_status?: string;
    approval_status?: EvaluationApprovalStatus;
    has_external_side_effect?: boolean;
    trace_incomplete?: boolean;
    retrieval_empty?: boolean;
    needs_review?: boolean;
    hard_gate_passed?: boolean;
}

interface Page<T> { items: T[]; total: number; limit?: number; offset?: number }

function withQuery(path: string, values: object): string {
    const params = new URLSearchParams();
    Object.entries(values).forEach(([key, value]) => { if (value) params.set(key, value); });
    const query = params.toString();
    return query ? `${path}?${query}` : path;
}

export const evaluationApi = {
    /** Loads server-owned one-click presets and owner-scoped baseline availability. */
    catalog: () => apiRequest<EvaluationCatalog>('/api/evaluations/catalog'),
    overview: () => apiRequest<EvaluationOverview>('/api/evaluations/overview'),
    trends: (filters: EvaluationTrendFilters = {}) => apiRequest<{ items: EvaluationTrendPoint[] }>(withQuery('/api/evaluations/metrics/trends', filters)),
    regressions: () => apiRequest<Page<EvaluationRegression>>('/api/evaluations/regressions'),
    datasets: () => apiRequest<Page<EvaluationDataset>>('/api/evaluations/datasets?limit=200&offset=0'),
    getDataset: (id: string) => apiRequest<EvaluationDatasetDetail>(`/api/evaluations/datasets/${id}`),
    createDataset: (payload: Record<string, unknown>) => apiRequest<EvaluationDataset>('/api/evaluations/datasets', { method: 'POST', body: JSON.stringify(payload) }),
    updateDatasetStatus: (id: string, status: 'annotating' | 'calibrated' | 'retired') => apiRequest<EvaluationDataset>(`/api/evaluations/datasets/${id}/status`, { method: 'POST', body: JSON.stringify({ status }) }),
    lockDataset: (id: string) => apiRequest<EvaluationDataset>(`/api/evaluations/datasets/${id}/lock`, { method: 'POST' }),
    suites: () => apiRequest<Page<EvaluationSuite>>('/api/evaluations/suites'),
    createSuite: (payload: Record<string, unknown>) => apiRequest<EvaluationSuite>('/api/evaluations/suites', { method: 'POST', body: JSON.stringify(payload) }),
    runs: () => apiRequest<Page<EvaluationRun>>('/api/evaluations/runs?limit=200&offset=0'),
    getRun: (id: string) => apiRequest<EvaluationRun & { cases: EvaluationCaseRun[] }>(`/api/evaluations/runs/${id}`),
    createRun: (payload: Record<string, unknown>) => apiRequest<EvaluationRun>('/api/evaluations/runs', { method: 'POST', body: JSON.stringify(payload), headers: { 'Idempotency-Key': crypto.randomUUID() } }),
    /** Starts a real evaluation while the backend keeps API keys only in encrypted AgentRun payload. */
    quickRun: (payload: EvaluationQuickRunRequest, idempotencyKey: string) => apiRequest<EvaluationRun>('/api/evaluations/quick-runs', {
        method: 'POST',
        body: JSON.stringify(payload),
        headers: { 'Idempotency-Key': idempotencyKey },
    }),
    cancelRun: (id: string) => apiRequest<Record<string, unknown>>(`/api/evaluations/runs/${id}/cancel`, { method: 'POST' }),
    retryRun: (id: string) => apiRequest<Record<string, unknown>>(`/api/evaluations/runs/${id}/retry-failed`, { method: 'POST' }),
    requestReview: (id: string, caseRunIds: string[] = [], failedOnly = true) => apiRequest<{ run_id: string; queued_count: number }>(`/api/evaluations/runs/${id}/request-review`, { method: 'POST', body: JSON.stringify({ case_run_ids: caseRunIds, failed_only: failedOnly }) }),
    caseRuns: (runId: string, filters: EvaluationCaseRunFilters = {}) => {
        const params = new URLSearchParams();
        for (const [key, value] of Object.entries(filters)) {
            if (value !== undefined && value !== null && value !== '') params.set(key, String(value));
        }
        const suffix = params.toString() ? `?${params.toString()}` : '';
        return apiRequest<Page<EvaluationCaseRun>>(`/api/evaluations/runs/${runId}/cases${suffix}`);
    },
    caseRun: (id: string) => apiRequest<EvaluationCaseRunDetail>(`/api/evaluations/case-runs/${id}`),
    annotationQueue: () => apiRequest<Page<EvaluationCaseRun & { agent_name: string }>>('/api/evaluations/annotations/queue'),
    annotations: (caseRunId: string) => apiRequest<Page<EvaluationAnnotation>>(`/api/evaluations/case-runs/${caseRunId}/annotations`),
    addAnnotation: (caseRunId: string, payload: Record<string, unknown>) => apiRequest<EvaluationAnnotation>(`/api/evaluations/case-runs/${caseRunId}/annotations`, { method: 'POST', body: JSON.stringify(payload) }),
    adjudicate: (annotationId: string, payload: Record<string, unknown>) => apiRequest<EvaluationAnnotation>(`/api/evaluations/annotations/${annotationId}/adjudicate`, { method: 'POST', body: JSON.stringify(payload) }),
    createCandidateDataset: (caseRunId: string, payload: Record<string, unknown>) => apiRequest<EvaluationDataset>(`/api/evaluations/case-runs/${caseRunId}/candidate-dataset`, { method: 'POST', body: JSON.stringify(payload) }),
    calibrations: () => apiRequest<Page<EvaluationCalibration>>('/api/evaluations/calibrations'),
    createCalibration: (payload: Record<string, unknown>) => apiRequest<EvaluationCalibration>('/api/evaluations/calibrations', { method: 'POST', body: JSON.stringify(payload) }),
    simulateCalibration: (id: string, payload: Record<string, unknown>) => apiRequest<Record<string, number>>(`/api/evaluations/calibrations/${id}/simulate`, { method: 'POST', body: JSON.stringify(payload) }),
    gates: () => apiRequest<Page<EvaluationGatePolicy>>('/api/evaluations/gates'),
    createGate: (payload: Record<string, unknown>) => apiRequest<EvaluationGatePolicy>('/api/evaluations/gates', { method: 'POST', body: JSON.stringify(payload) }),
    gateCheck: (runId: string, policyId?: string) => apiRequest<Record<string, unknown>>(`/api/evaluations/runs/${runId}/gate-check${policyId ? `?policy_id=${encodeURIComponent(policyId)}` : ''}`, { method: 'POST' }),
    traceLink: (agentRunId: string) => apiRequest<{ available: boolean; url: string | null; message: string | null }>(`/api/agent-runs/${agentRunId}/trace-link`),
};

/** Downloads an authenticated HTML report without exposing the user id in the URL. */
export async function downloadEvaluationReport(runId: string): Promise<void> {
    const response = await authFetch(`/api/evaluations/runs/${runId}/report?format=html`);
    if (!response.ok) throw new Error(`报告导出失败: HTTP ${response.status}`);
    const blob = await response.blob();
    const link = document.createElement('a');
    link.href = URL.createObjectURL(blob);
    link.download = `evaluation-${runId}.html`;
    link.click();
    URL.revokeObjectURL(link.href);
}
